### ISSUE TITLE
JsInterop and ClassCastException when using Optional and Integer

### ISSUE THREAD

**Issue by Gebon** [gwtproject/gwt#9621]
**GWT version:** 2.8.1
**Browser (with version):** Google Chrome 67.0.3396.99 (Official Build) (64-bit)
**Operating System:** Ubuntu 18.04

---

##### Description

ClassCastException is raised, when I'm trying to assign the resutl of `Optional.ofNullable(referenceInteger).orElse(0)` to a variable of type `Integer` and `referenceInteger` is not null

##### Steps to reproduce

```java
package example;

import jsinterop.annotations.JsMethod;
import java.util.Optional;

public class Example {
  @JsMethod
  public void example(Integer i) {
    Integer result = Optional.ofNullable(i).orElse(0);
  }
}
```

And in JS:
```js
var ex = new example.Example();
ex.example(); // fine
ex.example(0); // fails
```

##### Question

Is this the desired behavior? How could I make not required function arguments with Integer?

**Comment by Gebon** [comment]
FYI: I've found a workaround

```java
package example;

import jsinterop.annotations.JsMethod;
import java.util.Optional;

public class Example {
  @JsMethod
  public void example(Integer i) {
    Number result = Optional.ofNullable(i).orElse(0);
    Integer intResult = result.intValue();
  }
}
```

**Comment by niloc132** [comment]
`Integer` is not a valid type to expose to JS - I'm surprised you didn't see a compiler warning, or did you and ignored it?

The primitive `int` can be passed in safely, but isn't nullable. In JS, the Number type matches only the java boxed (nullable) `Double` type - you could make that your param type if you wanted to support nulls, and then coerce/cast to int if non-null (and then box). Keep in mind that a user might call your Integer method with `0.5` and it won't be checked or coerced at all by GWT, the compiler implicitly trusts the types you say will be passed in.

In other words: yes, this is the expected behavior, since the boxed type Integer isn't compatible with JS.

**Comment by Gebon** [comment]
Thank you for an answer! Now it's much more clear to me
